App crashes when back button is pressed while data being fetched

**Actual Behaviour**

The app crashes when the back button is pressed in between the process of fetching data from PSLab device in Logic Analyzer Instrument

**Expected Behaviour**

The app shouldn't crash

**Steps to reproduce it**

Go to Logic Analyzer and press back button when loading is shown on the screen

**LogCat for the issue**

`E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #1
                  Process: org.fossasia.pslab, PID: 10180
                  java.lang.RuntimeException: An error occurred while executing doInBackground()
                      at android.os.AsyncTask$3.done(AsyncTask.java:353)
                      at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
                      at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
                      at java.util.concurrent.FutureTask.run(FutureTask.java:271)
                      at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
                      at java.lang.Thread.run(Thread.java:764)
                   Caused by: java.lang.IllegalStateException: Fragment LALogicLinesFragment{4ae62fc} not attached to a context.
                      at android.support.v4.app.Fragment.requireContext(Fragment.java:614)
                      at android.support.v4.app.Fragment.getResources(Fragment.java:678)
                      at org.fossasia.pslab.fragment.LALogicLinesFragment$CaptureOne.doInBackground(LALogicLinesFragment.java:562)
                      at org.fossasia.pslab.fragment.LALogicLinesFragment$CaptureOne.doInBackground(LALogicLinesFragment.java:539)
                      at android.os.AsyncTask$2.call(AsyncTask.java:333)
                      at java.util.concurrent.FutureTask.run(FutureTask.java:266)
                      at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245) 
                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
                      at java.lang.Thread.run(Thread.java:764) `

**Would you like to work on the issue?**

Yes

--- a/app/src/main/java/org/fossasia/pslab/activity/LogicalAnalyzerActivity.java
+++ b/app/src/main/java/org/fossasia/pslab/activity/LogicalAnalyzerActivity.java
@@ -2,7 +2,6 @@ package org.fossasia.pslab.activity;
 
 import android.os.Bundle;
 import android.support.annotation.Nullable;
-import android.support.v4.app.FragmentTransaction;
 import android.support.v7.app.AppCompatActivity;
 import android.support.v7.widget.Toolbar;
 import android.view.MenuItem;
@@ -24,6 +23,7 @@ public class LogicalAnalyzerActivity extends AppCompatActivity {
     @BindView(R.id.logical_analyzer_toolbar)
     Toolbar toolbar;
     private ScienceLab scienceLab;
+    private boolean isRunning = false;
 
     @Override
     protected void onCreate(@Nullable Bundle savedInstanceState) {
@@ -50,6 +50,11 @@ public class LogicalAnalyzerActivity extends AppCompatActivity {
 
     @Override
     public void onBackPressed() {
-        finish();
+        if (!isRunning)
+            finish();
+    }
+
+    public void setStatus(boolean status) {
+        isRunning = status;
     }
 }

--- a/app/src/main/java/org/fossasia/pslab/fragment/LALogicLinesFragment.java
+++ b/app/src/main/java/org/fossasia/pslab/fragment/LALogicLinesFragment.java
@@ -39,6 +39,7 @@ import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;
 import com.github.mikephil.charting.listener.OnChartValueSelectedListener;
 
 import org.fossasia.pslab.R;
+import org.fossasia.pslab.activity.LogicalAnalyzerActivity;
 import org.fossasia.pslab.communication.ScienceLab;
 import org.fossasia.pslab.communication.digitalChannel.DigitalChannel;
 import org.fossasia.pslab.others.ChannelAxisFormatter;
@@ -179,6 +180,7 @@ public class LALogicLinesFragment extends Fragment {
         xCoordinateText.setText("Time:  0.0 mS");
         progressBar = v.findViewById(R.id.la_progressBar);
         progressBar.setVisibility(View.GONE);
+        ((LogicalAnalyzerActivity) getActivity()).setStatus(false);
 
         // Bottom Sheet guide
         bottomSheet = (LinearLayout) v.findViewById(R.id.bottom_sheet);
@@ -348,6 +350,7 @@ public class LALogicLinesFragment extends Fragment {
 
                         if (channelMode == 1) {
                             progressBar.setVisibility(View.VISIBLE);
+                            ((LogicalAnalyzerActivity) getActivity()).setStatus(true);
                             Thread monitor = new Thread(new Runnable() {
                                 @Override
                                 public void run() {
@@ -594,6 +597,7 @@ public class LALogicLinesFragment extends Fragment {
                 // Plot the fetched data
                 updateLogicLines(xaxis, yaxis);
                 progressBar.setVisibility(View.GONE);
+                ((LogicalAnalyzerActivity) getActivity()).setStatus(false);
 
                 List<ILineDataSet> dataSets = new ArrayList<>();
                 LineDataSet lineDataSet = new LineDataSet(tempInput, channelNames.get(0));
@@ -621,6 +625,7 @@ public class LALogicLinesFragment extends Fragment {
                 }
             } else {
                 progressBar.setVisibility(View.GONE);
+                ((LogicalAnalyzerActivity) getActivity()).setStatus(false);
                 Toast.makeText(getContext(), getResources().getString(R.string.no_data_generated), Toast.LENGTH_SHORT).show();
             }
         }

